File: problems/leetcode/0200/285.cpp
// 285. Inorder Successor in BST
// https://leetcode.com/problems/inorder-successor-in-bst/

using namespace std;

#include <cstddef>

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
  TreeNode *inorderSuccessor(TreeNode *root, TreeNode *p) {
    bool isSuccessor = false;
    return inorderTraversal(root, p, isSuccessor);
  }

  TreeNode *inorderTraversal(TreeNode *node, TreeNode *p, bool &isSuccessor) {
    if (node->left) {
      TreeNode *left = inorderTraversal(node->left, p, isSuccessor);
      if (left) {
        return left;
      }
    }

    if (isSuccessor) {
      return node;
      isSuccessor = false;
    }

    if (node == p) {
      isSuccessor = true;
    }

    if (node->right) {
      TreeNode *right = inorderTraversal(node->right, p, isSuccessor);
      if (right) {
        return right;
      }
    }

    return nullptr;
  }
};

File: problems/leetcode/0200/285/285.py
# 285. Inorder Successor in BST
# https://leetcode.com/problems/inorder-successor-in-bst/description/

from typing import *

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def inorderSuccessor(self, root: TreeNode, p: TreeNode) -> Optional[TreeNode]:
        return self.traverseInorderSuccessor(root, p, None)

    def traverseInorderSuccessor(self, node: TreeNode, p: TreeNode, successor: Optional[TreeNode]) -> Optional[TreeNode]:
        if p.val < node.val:
            return self.traverseInorderSuccessor(node.left, p, node)
        elif node.val < p.val:
            return self.traverseInorderSuccessor(node.right, p, successor)
        elif node.right:
            return self.traverseSmallest(node.right)
        else:
            return successor

    def traverseSmallest(self, node: TreeNode) -> TreeNode:
        if node.left:
            return self.traverseSmallest(node.left)
        else:
            return node
